File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.java
import java.util.*;

class Solution {
    public int solution(int[][] jobs) {
        int answer = 0;
        PriorityQueue<int[]> heap = new PriorityQueue<>((o1, o2) -> o1[1] - o2[1]);
        int currentTime = 0;
        int num = jobs.length;
        Arrays.sort(jobs, (o1, o2) -> o1[0] - o2[0]);
        int k = 0;
        int index = 0;
        while (k < num){
            while (index < num && jobs[index][0] <= currentTime){
                heap.add(jobs[index++]);
            }
            
            if (!heap.isEmpty()){
                int[] currentJob = heap.poll();
                currentTime += currentJob[1];
                answer += currentTime - currentJob[0];
                k++;
            } else{
                currentTime++;
            }
        } 
        return answer/num;
    }
}

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.py
import heapq
def solution(jobs):
    answer = 0
    heap = []
    current_time = 0
    num = len(jobs)
    start = -1
    jobs.sort()
    k=0
    while k < num:
        temp = jobs[:]
        for job in temp:
            if job[0] <= current_time:
                heapq.heappush(heap,[job[1],job[0]])
                jobs.pop(0)
        if len(heap) > 0:
            current_job = heapq.heappop(heap)
            current_time += current_job[0]
            answer += current_time - current_job[1]
            k += 1
        else:
            current_time += 1
    return answer//num